### PR TITLE
Implement scheduler using tdd principles

### PR DESCRIPTION
**Pull request by jacoblipech**

### CODE REVIEW

File: walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 1]
```diff
@@ -14,10 +14,85 @@
...
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+
+    // corner case: requested duration more than the day duration
+    if (reqDuration > TimeRange.WHOLE_DAY.duration()) {
+      return Arrays.asList();
+    }
+
+    // check if each event's attendees clash with the request attendees sorted by starting time
+    List<TimeRange> eventTimes = events.stream()
```
Comment (123cristen) [inline walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:45]: Nice use of stream!

File: walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 2]
```diff
@@ -14,10 +14,85 @@
...
+  private void addPossibleTime(int start, int end, long reqDuration) {
+    if (end - start >= reqDuration) {
+      possibleMeetingTime.add(TimeRange.fromStartEnd(start, end, false));
+    }
+  }
+
+  /**
+   * Checks to ensure that events are added only when attendee overlaps.
+   */
+  private Boolean isAffectedEvent(Set<String> eventAttendees, Collection<String> reqAttendees) {
```
Comment (123cristen) [inline walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Consider renaming to be more descriptive, I.e. `eventContainsRequiredAttendee`. Also, with a name like that it'd be more clear to pass in the event object itself rather than the eventAttendees :)

File: walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 3]
```diff
@@ -14,10 +14,85 @@
...
+    * Adds a possible meeting time to the query
+    */
+  private void addPossibleTime(int start, int end, long reqDuration) {
+    if (end - start >= reqDuration) {
+      possibleMeetingTime.add(TimeRange.fromStartEnd(start, end, false));
+    }
+  }
+
+  /**
+   * Checks to ensure that events are added only when attendee overlaps.
```
Comment (123cristen) [inline walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: The comment should also just describe what the method does, not the purpose of it - so maybe rephrase ` Checks whether an event contains a required attendee.`

File: walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 4]
```diff
@@ -14,10 +14,85 @@
...
+      int nextEventStart = eventTimes.get(i).start();
+      int prevEventEnd = eventTimes.get(i-1).end();
+      int currEventEnd = eventTimes.get(i).end();
+
+      addPossibleTime(possibleStartTime, nextEventStart, reqDuration);
+      // ensures that the next free time is the based on the latest event ending time
+      possibleStartTime = Math.max(prevEventEnd, currEventEnd);
+    }
+
+    addPossibleTime(possibleStartTime, TimeRange.END_OF_DAY + 1, reqDuration);
```
Comment (123cristen) [inline walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:75]: Add a comment explaining why you have to do this one last time?

File: walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 5]
```diff
@@ -14,10 +14,85 @@
...
+                                      .collect(Collectors.toList());
+
+    // no event affecting required attendees
+    if (eventTimes.isEmpty()) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+
+    int possibleStartTime = TimeRange.START_OF_DAY;
+
+    // check first event starting time
```
Comment (123cristen) [inline walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: In general, I'd avoid comments that just state exactly what is being done, since it's self evident from the code. (Except sometimes for class or method javadoc comments to give an overview). However, it would be helpful to have a comment in this case like maybe `// Consider the time window between the start of the day and the first event.`

**Review by 123cristen — CHANGES_REQUESTED** [review]
Nice! Looks very clean overall, great use of helper methods! Just a few comments on commenting style + naming.

File: walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 6]
```diff
@@ -14,10 +14,85 @@
...
+    if (end - start >= reqDuration) {
+      possibleMeetingTime.add(TimeRange.fromStartEnd(start, end, false));
+    }
+  }
+
+  /**
+   * Checks to ensure that events are added only when attendee overlaps.
+   */
+  private Boolean isAffectedEvent(Set<String> eventAttendees, Collection<String> reqAttendees) {
+    for (String attendee : reqAttendees) {
```
Comment (123cristen) [inline walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: This logic can be simplified using `Collections.disjoint()` :)

Comment (jacoblipech) [reply walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: worh works wonders! :) thanks for sharing!

**Review by 123cristen — APPROVED** [review]
Looks great!
